edx_exams/apps/lti/signals/tests/test_handlers.py

```diff
@@ -0,0 +1,142 @@
+"""
+Test signal handlers.
+"""
+import uuid
+from unittest.mock import patch
+
+from django.test import TestCase
+
+from edx_exams.apps.api.test_utils.factories import UserFactory
+from edx_exams.apps.core.models import Exam, ExamAttempt, ProctoringProvider
+from edx_exams.apps.core.statuses import ExamAttemptStatus
+from edx_exams.apps.lti.signals.handlers import assessment_started
+
+
+class TestAssessmentStarted(TestCase):
+    """
+    Test the assessment_started signal handler.
+    """
+
+    def setUp(self):
+        super().setUp()
+
+        self.user = UserFactory(username='test', is_active=True, is_staff=False)
+        self.user.set_password('password')
+        self.user.save()
+
+        self.test_provider = ProctoringProvider.objects.create(
+            name='test_provider',
+            verbose_name='testing provider',
+            lti_configuration_id='123456789',
+            tech_support_phone='1118976309',
+            tech_support_email='test@example.com',
+        )
+
+        self.course_id = 'course-v1:edx+test+f19'
+        self.content_id = '11111111'
+
+        self.exam = Exam.objects.create(
+            resource_id=str(uuid.uuid4()),
+            course_id=self.course_id,
+            provider=self.test_provider,
+            content_id=self.content_id,
+            exam_name='test_exam',
+            exam_type='proctored',
+            time_limit_mins=30,
+            due_date='2021-07-01 00:00:00',
+            hide_after_due=False,
+            is_active=True
+        )
+
+        self.attempt = ExamAttempt.objects.create(
+            user=self.user,
+            exam=self.exam,
+            attempt_number=1111111,
+            status=ExamAttemptStatus.created,
+            start_time=None,
+            allowed_time_limit_mins=None,
+        )
+
+    def test_assessment_started_updated_attempt(self):
```

Comment:
Is there a reason that `update_attempt_status` isn't being mocked here? That function already has tests, so instead of testing the whole code path here you could check that the mock has been called

Reply:
I don't typically mock unless it's absolutely necessary. I had to mock in the some of these other tests because I'm testing that nothing happened, and that's hard to do without asserting on _something_, so I chose to ensure that `update_attempt_status` wasn't called. I'm open to alternatives to avoid mocking in these tests!

Otherwise, mocking out that function feels like testing an implementation detail. I'd like to test the result of calling the handler, not necessarily how the handler achieves that result. In my opinion, the test shouldn't care how the status was updated - just that it was updated somehow. If I mock out `update_attempt_status`, then these tests are reliant on the assumption that `update_attempt_status` is called in the function, and they will break anytime we change the way we update attempt statuses or change the function name, making them brittle. If we don't mock out the function, the tests will continue to function regardless.

What do you think?

Reply:
That makes sense!